### PR TITLE
Make the recursive option of cargo-sweep recurse past Cargo directories

### PR DESCRIPTION
**Pull request by bes**
When passing the -r (recursive) option to cargo-sweep, it is supposed to be recursive over the subdirectories of the root path it is given (whether implicit or explicit). cargo-sweep erronously assumed that a Cargo project could not contain unrelated subprojects.

This patch lets cargo-sweep keep recursing past cargo roots.

The drawback of this patch is that cargo-sweep will need to recurse more, and as such becomes a bit slower to run.

Fixes https://github.com/holmgr/cargo-sweep/issues/66

### CODE REVIEW

**Review by jyn514 — COMMENTED** [review]
Can you please add a test for the new behavior? You may need to add the project from https://github.com/holmgr/cargo-sweep/issues/66 into `tests`, see `tests/integration.rs` for some existing code you can use as an example.

Otherwise this looks fine to me :) the CI failures are unrelated and should be fixed if you rebase over master.

**Comment by bes** [comment]
@jyn514 I finally got some time to write an integration test, and I think it came out OK.

But I have to say, the state of the `integration.rs` file is not very good. I tried to improve it, but I have some... feelings.

Tests should be easy to follow step-by-step by reading the test function, but folks have tried to be too clever with the code. Deduplication, DRY, and utility functions that assume too much about how a test works will end up hurting tests as much as they help improve regular code.

I had to start by documenting a lot of the code that I didn't understand and then patch my integration test on top of that.

Furthermore, the `empty_project_output` didn't run locally for me. Either I broke it somehow, or the regexp was written by someone with another type of machine than me where the file system outputs files and directories in a different order. We'll see once someone approves my PR for testing. My recommendation would be to fix that test to be platform agnostic (unless I broke it, of course).

**Comment by bes** [comment]
Yeah I just confirmed that the test `empty_project_output` is broken on the `master` branch on my machine as well (2019 intel mac)

**Comment by bes** [comment]
The regex requires this order

```rust
        \[DEBUG\] Successfully removed: ".+libsample_project-.+\.rlib"
        \[DEBUG\] Successfully removed: ".+libsample_project-.+\.rmeta"
        \[DEBUG\] Successfully removed: ".+sample_project-.+\.d"
        \[DEBUG\] Successfully removed: ".+.fingerprint.+sample-project-.+"
```

But my machine outputs this order

```shell
[DEBUG] Successfully removed: "/var/folders/g8/qgn3gbf514b_p34n_jc_1_lc0000gn/T/.tmpKdJQnA/debug/deps/sample_project-f27e05aa5c13ed9b.d"
[DEBUG] Successfully removed: "/var/folders/g8/qgn3gbf514b_p34n_jc_1_lc0000gn/T/.tmpKdJQnA/debug/deps/libsample_project-f27e05aa5c13ed9b.rlib"
[DEBUG] Successfully removed: "/var/folders/g8/qgn3gbf514b_p34n_jc_1_lc0000gn/T/.tmpKdJQnA/debug/deps/libsample_project-f27e05aa5c13ed9b.rmeta"
[DEBUG] Successfully removed: "/var/folders/g8/qgn3gbf514b_p34n_jc_1_lc0000gn/T/.tmpKdJQnA/debug/.fingerprint/sample-project-f27e05aa5c13ed9b"
```

**Comment by bes** [comment]
I managed to fix it by changing the regexp to this monstrosity. My guess is that regex isn't the right tool for the job, though.

```rust
        (\[DEBUG\] Successfully removed: (".+libsample_project-.+\.rlib"|".+libsample_project-.+\.rmeta"|".+sample_project-.+\.d"|".+.fingerprint.+sample-project-.+")\n{0,1}){4}
```

File: tests/integration.rs [hunk 1]
```diff
@@ -57,10 +60,7 @@ fn sweep(args: &[&str]) -> Command {
 }
 
 /// Sets the target folder and runs the given command
-fn run<'a>(mut cmd: impl BorrowMut<Command>, target: impl Into<Option<&'a Path>>) -> Assert {
-    if let Some(target) = target.into() {
-        cmd.borrow_mut().env("CARGO_TARGET_DIR", target);
-    }
```
Comment (marcospb19) [inline tests/integration.rs:63]: > Tests should be easy to follow step-by-step by reading the test function, but folks have tried to be too clever with the code.

I wasn't "trying to be clever", `Into<Option<T>>` is just some way for you to provide an argument without `Some()` :smiley:.

---

Also, I don't see any clear reason for reverting this, the `golden_reference` test is the only one that doesn't require a `target`, because it returns at `--help` and `--version` (nothing is done), so it makes sense to explicitly say "This runs with no target". All other tests expect some target path so you just provide it.

But well, it's just a style change so it kinda doesn't matter what I think about it.

File: tests/integration.rs [hunk 2]
```diff
@@ -221,10 +256,7 @@ fn empty_project_output() -> TestResult {
             \),
         \]
         \[DEBUG\] cleaning: ".+debug" with remove_not_built_with_in_a_profile
-        \[DEBUG\] Successfully removed: ".+libsample_project-.+\.rlib"
-        \[DEBUG\] Successfully removed: ".+libsample_project-.+\.rmeta"
-        \[DEBUG\] Successfully removed: ".+sample_project-.+\.d"
-        \[DEBUG\] Successfully removed: ".+.fingerprint.+sample-project-.+"
+        (\[DEBUG\] Successfully removed: (".+libsample_project-.+\.rlib"|".+libsample_project-.+\.rmeta"|".+sample_project-.+\.d"|".+.fingerprint.+sample-project-.+")\n{0,1}){4}
```
Comment (marcospb19) [inline tests/integration.rs:259]: I agree that regex is kinda bad for this use-case, but I didn't know what to use instead, do you have any suggestions?

I thought I had fixed the order problem now several times, but it struck again.

Maybe this should just expect 4 lines?
```txt
        \[DEBUG\] Successfully removed: ".+sample.+"
        \[DEBUG\] Successfully removed: ".+sample.+"
        \[DEBUG\] Successfully removed: ".+sample.+"
        \[DEBUG\] Successfully removed: ".+sample.+"
```
or
```
(        \[DEBUG\] Successfully removed: ".+sample.+"){4}
```
What do you think? The line you added is a bit weird to read ("weird to read" is almost ironic in a RegEx context).

Comment (marcospb19) [reply tests/integration.rs:259]: Here's something terrible nobody asked:
```regex
(\[DEBUG\] Successfully removed: .+((lib)?sample).+(rlib|rmeta|d).+){4}
```

File: tests/integration.rs (hunk 1 above)
Comment (bes) [reply tests/integration.rs:63]: Hi @marcospb19 I wasn't primarily thinking about the `run` function, but more the whole setup of the tests. For example, the fact that the test modifies the `CARGO_TARGET_DIR` at all is weird to me.  I didn't want to criticize anyone in particular, sorry 🙏

My tests instead copy the whole project into a temp folder and runs it there, which gives the following advantages:

* It isn't necessary to give the `target` directory at all.
* Tests are more realistic, with regards to how `cargo-sweep` actually works.
* Bugs that affect the project files can be found.

But to answer your specific question, I'd say that the `run` function is an abstraction over running _something_, not necessarily the `sweep` command, so including the target directory as a parameter to it would make it weird if someone would like to use `run` to run other commands.

One good example is the `build` function that also just assumes that the build must put target files in a temporary directory, which is so weird to me. It was used in too many places so I didn't want to change it.

Overall I'd say that my feeling is that if all tests were changed to use the same test methodology `recursive_multiple_root_workspaces` (which doesn't use the `build` function), they would be easier to read, understand, and also later modify when future requirements change.

In short: Tests _should be verbose_ 🥰

File: tests/integration.rs (hunk 2 above)
Comment (bes) [reply tests/integration.rs:259]: @marcospb19 Let's just take the first suggestion? =D

**Review by marcospb19 — APPROVED** [review]
LGTM, sorry my last PR gave you some conflicts xD.

These should be easy to fix because you only changed one line at `main.rs`, and there aren't many changes on `integration.rs`, here's [the diff](https://github.com/holmgr/cargo-sweep/pull/87/files), just in case it helps.

File: tests/integration.rs (hunk 1 above)
Comment (jyn514) [reply tests/integration.rs:63]: > For example, the fact that the test modifies the CARGO_TARGET_DIR at all is weird to me.

Using a temporary directory seems okay, but in that case please explicitly unset it so the test still pass if I have CARGO_TARGET_DIR set locally.

**Comment by bes** [comment]
Updated

File: tests/integration.rs (hunk 1 above)
Comment (bes) [reply tests/integration.rs:62]: Thanks @jyn514 I have updated the `recursive_multiple_root_workspaces` test to ignore `CARGO_TARGET_DIR`

**Comment by bes** [comment]
Looks like there is a "bug" in `format_bytes(bytes: u64)`, it doesn't agree with the values from the `human-size` crate?

`human-size` expects bytes to be suffixed `B` but `format_bytes` outputs the string `"bytes"`.

File: tests/integration.rs [hunk 3]
```diff
@@ -341,3 +372,89 @@ fn subcommand_usage() -> TestResult {
...
+/// * Do a dry run sweep, asserting that nothing was actually cleaned.
+/// * Do a proper sweep, asserting that the target directory size is back down to its
+///   expected size.
+///
+/// Please note that there is some ceremony involved, namely correctly setting:
+/// * `CARGO_TARGET_DIR` to `unset`, so as not to clash with users who use `CARGO_TARGET_DIR`
+///   when they invoke `cargo test` for running these tests.
+/// * `CARGO_INCREMENTAL` to `0`, because cargo-sweep doesn't yet clear incremental files.
+#[test]
+fn recursive_multiple_root_workspaces() -> TestResult {
```
Comment (jyn514) [inline tests/integration.rs:397]: This test is not effective, it passes even without your change.
```
; git log -n1
commit 8ca98385c1d7af79b491a30eb330f12a4b7173e2 (HEAD)
Author: Erik Zivkovic <erik@zivkovic.se>
Date:   Wed Feb 1 23:13:52 2023 +0100

    Make cargo-sweep output parselable by human-size
; git status 
HEAD detached at FETCH_HEAD
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   src/main.rs
        modified:   src/util.rs

; git diff origin/master src
; cargo test recursive -q

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 3 filtered out; finished in 0.00s


running 1 test
.
test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 9 filtered out; finished in 1.03s
```

File: tests/integration.rs [hunk 4]
```diff
@@ -341,3 +372,89 @@ fn subcommand_usage() -> TestResult {
...
+    let target = tempdir()?;
```
Comment (jyn514) [inline tests/integration.rs:398]: This is ambiguous with the `target/` directory cargo generates - please call this `temp_workspace_dir` or something like that.

File: tests/integration.rs [hunk 5]
```diff
@@ -341,3 +372,89 @@ fn subcommand_usage() -> TestResult {
...
+
+    // Build workspace crates
+    run(Command::new(env!("CARGO"))
+        // If someone has built & run these tests with CARGO_TARGET_DIR,
+        // we need to override that.
+        .env_remove("CARGO_TARGET_DIR")
+        // Don't output incremental build artifacts
+        .env("CARGO_INCREMENTAL", "0")
+        .arg("build")
+        .current_dir(target.path().join("nested-root-workspace")));
```
Comment (jyn514) [inline tests/integration.rs:430]: nit: Please deduplicate these `cargo` commands - you can change the existing `cargo` command to require an absolute path instead of a relative path as the `project` argument.

File: tests/integration.rs [hunk 6]
```diff
@@ -341,3 +372,89 @@ fn subcommand_usage() -> TestResult {
...
+
+    let old_size = get_size(target.path())?;
+
+    // Build bin-crate
+    run(Command::new(env!("CARGO"))
+        // If someone has built & run these tests with CARGO_TARGET_DIR,
+        // we need to override that.
+        .env_remove("CARGO_TARGET_DIR")
+        // Don't output incremental build artifacts
+        .env("CARGO_INCREMENTAL", "0")
```
Comment (jyn514) [inline tests/integration.rs:415]: This isn't actually necessary, we aren't asserting that the size of the cleaned target dir is 0. You can just leave the artifacts be and they'll get removed automatically when the temporary directory is deleted.

**Review by jyn514 — COMMENTED** [review]
The source changes seem ok, but I had a few comments about the test changes - most of them are non-blocking except the fact that the test isn't actually effective.

File: tests/integration.rs (hunk 3 above)
Comment (bes) [reply tests/integration.rs:397]: I agree that the last change `8ca98385c1d7af79b491a30eb330f12a4b7173e2` doesn't do anything locally, but as you can see it failed CI because for some reason that windows machine generated a line that looks like this:

```
[INFO] Would clean: 431.00 bytes from "C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\.tmpusn7fA\\nested-root-workspace\\bin-crate\\target"
```

See output of this test run
https://github.com/holmgr/cargo-sweep/actions/runs/4062073908/jobs/7008162492

File: tests/integration.rs (hunk 6 above)
Comment (bes) [reply tests/integration.rs:415]: Thanks, I will remove these lines.

File: tests/integration.rs (hunk 5 above)
Comment (bes) [reply tests/integration.rs:430]: Done

File: tests/integration.rs (hunk 4 above)
Comment (bes) [reply tests/integration.rs:398]: Done

File: tests/integration.rs (hunk 3 above)
Comment (jyn514) [reply tests/integration.rs:366]: You're misunderstanding, sorry - I'm not saying 8ca98385c1d7af79b491a30eb330f12a4b7173e2 doesn't do anything, I'm saying "this test doesn't test anything". It passes both before and after all your changes to `src/`.

https://github.com/holmgr/cargo-sweep/actions/runs/4062073908/jobs/7008162492 looks like a difference in the size of the target directory, it makes sense to me that 8ca98385c1d7af79b491a30eb330f12a4b7173e2 fixes it.

Comment (bes) [reply tests/integration.rs:366]: Gotcha, I understand now. Done.

**Comment by bes** [comment]
> The source changes seem ok, but I had a few comments about the test changes - most of them are non-blocking except the fact that the test isn't actually effective.

Fixed 👍

File: tests/integration.rs [hunk 7]
```diff
@@ -64,17 +64,29 @@ fn run<'a>(mut cmd: impl BorrowMut<Command>, target: impl Into<Option<&'a Path>>
     assert
 }
 
-/// Returns the size of the build directory.
+/// Runs a cargo build of the named project (which is expected to exist as a direct
+/// child of the `tests` directory.
```
Comment (jyn514) [inline tests/integration.rs:68]:
```suggestion
/// child of the `tests` directory).
```

File: tests/integration.rs [hunk 8]
```diff
@@ -310,3 +341,117 @@ fn usage() -> TestResult {
...
+///
+/// Step-by-step:
+///
+/// * Create a new temporary directory. Other tests only output the built files into
+///   the target directory, but this test copies the whole template project into the
+///   target directory.
+/// * Build the two different (nested) projects using cargo build, but without incremental
+///   compilation, because cargo-sweep doesn't remove files built by incremental compilation,
+///   which affects this test since this test doesn't use a separate target folder for just
+///   the target files.
```
Comment (jyn514) [inline tests/integration.rs:356]: This comment is outdated.
```suggestion
/// * Build the two different (nested) projects using cargo build
```

File: tests/integration.rs [hunk 9]
```diff
@@ -310,3 +341,117 @@ fn usage() -> TestResult {
...
+/// * Do a dry run sweep, asserting that nothing was actually cleaned.
+/// * Do a proper sweep, asserting that the target directory size is back down to its
+///   expected size.
+///
+/// Please note that there is some ceremony involved, namely correctly setting:
+/// * `CARGO_TARGET_DIR` to `unset`, so as not to clash with users who use `CARGO_TARGET_DIR`
+///   when they invoke `cargo test` for running these tests.
+/// * `CARGO_INCREMENTAL` to `0`, because cargo-sweep doesn't yet clear incremental files.
```
Comment (jyn514) [inline tests/integration.rs:364]:
```suggestion
```

File: tests/integration.rs [hunk 10]
```diff
@@ -310,3 +341,117 @@ fn usage() -> TestResult {
...
+        // If someone has built & run these tests with CARGO_TARGET_DIR,
+        // we need to override that.
+        cmd.env_remove("CARGO_TARGET_DIR")
+            .current_dir(temp_workspace_dir.path())
+    })?;
+    assert_sweeped_size(temp_workspace_dir.path(), actual_cleaned, final_build_size)?;
+    assert_eq!(actual_cleaned, expected_cleaned);
+
+    // Measure the size of the nested root workspace and the bin crate after cargo-sweep is invoked,
+    // and assert that both of their sizes have been reduced.
```
Comment (jyn514) [inline tests/integration.rs:442]: It took me a second to realize why this assertion passed.
```suggestion
    // and assert that both of their sizes have been reduced.
    // This works because by default cargo generates the `target/` directory in a subdirectory of the source dir.
```

**Review by jyn514 — APPROVED** [review]
Thanks! This looks really good :) I just have a couple nits.

File: tests/integration.rs (hunk 7 above)
Comment (bes) [reply tests/integration.rs:68]: Done

File: tests/integration.rs (hunk 8 above)
Comment (bes) [reply tests/integration.rs:356]: Done

File: tests/integration.rs (hunk 9 above)
Comment (bes) [reply tests/integration.rs:364]: Done

File: tests/integration.rs (hunk 10 above)
Comment (bes) [reply tests/integration.rs:442]: Counter-nit: `target/` doesn't end up in the "source dir", it ends up in the "package root" :)

Done

**Comment by jyn514** [comment]
Thansk for sticking with this!
